### PR TITLE
[RFC] high-level dispatching transform pipeline

### PR DESCRIPTION
**Pull request by vmoens**
I propose a high-level dispatching of the transfom, such that `transform.forward` only sees tensor(-like) inputs.
The `TransformDispatch` class should be used as a wrapper for a transform (or a composition of transforms). Provided that those are Transform instances, then TransformDispatch will take care of keeping the random parameters in memory for the time that the transform is applied. 
The main advantage i see is code clarity: the transform forward methods only sees a single data type, whereas other implementation have tedious routing scripts and methods. They also can receive and return a wide variety of datatypes, which -- I think -- is not very neat.
I have not proposed a method to wrap the transforms in TransformDispatch in a training script -- i'm open to suggestions on that one. I guess it will strongly depend on what the vision datasets will look like in the future.

### CODE REVIEW

File: torchvision/transforms/_core.py [hunk 1]
```diff
@@ -23,17 +24,37 @@ def query_sample(sample: Any, fn: Callable[[Any], Optional[T]]) -> Iterator[T]:
             yield result
 
 
+class TransformDispatch():
+    def __init__(cls, transforms):
+        cls.transforms = transforms
+        cls.transforms.set_reset_auto(False)
```
Comment (pmeier) [inline torchvision/transforms/_core.py]: Not sure why you renamed `self` to `cls` here, since the value will be an actual instance rather than a class. If you want the class, you need to decorate the method with `@classmethod`. Plus, some dunder method are implicitly class methods such as `__init_subclass__`. 

For everything that I see below, you most likely do not want to use class methods and attributes as they will be the same for every object of that class.

**Review by pmeier — COMMENTED** [review]
This is a valid approach, but so far it is hard to judge how the UX/DX would look like. Would you please provide a proof-of-concept implementation, so it can be compared to my proposal in #1? 

---

> The main advantage i see is code clarity: the transform forward methods only sees a single data type, whereas other implementation have tedious routing scripts and methods. 

I don't understand. How would you solve the fact that we need different transformations based on the feature type? Will every transformation implement their own dispatch in the forward method? If so, how would it look like?

> They also can receive and return a wide variety of datatypes, which -- I think -- is not very neat.

That is true, but somewhat needed. The new datasets will return a dictionary of possibly nested features and we still want to keep BC and thus have to also accept regular tensors. 

Plus, in #1 I've gone to great length to hide the dispatch from someone who is writing a new transform. They only need to write the actual transform, which I conjecture is what most people will actually care about. So the extra complexity should only be on us, the maintainers, but I'm willing to go there to have a very clean UX/DX.

> I have not proposed a method to wrap the transforms in TransformDispatch in a training script -- i'm open to suggestions on that one. I guess it will strongly depend on what the vision datasets will look like in the future.

There is the `transforms_poc.py` script that contains the most basic use cases. Could you adapt your implementation or the script itself to make it runnable?

File: torchvision/transforms/_core.py (hunk 1 above)
Comment (vmoens) [reply torchvision/transforms/_core.py]: sorry about that it's a mistake from my part, will fix it

**Comment by vmoens** [comment]
> I don't understand. How would you solve the fact that we need different transformations based on the feature type? Will every transformation implement their own dispatch in the forward method? If so, how would it look like?

Exactly like you did, i think the code is transparent on this regard. The only thing that changes is that you can only pass tensor(-like) objects to the transforms, not dictionaries or lists or anything else. 

> Plus, in #1 I've gone to great length to hide the dispatch from someone who is writing a new transform. They only need to write the actual transform, which I conjecture is what most people will actually care about. So the extra complexity should only be on us, the maintainers, but I'm willing to go there to have a very clean UX/DX.

And that stays exactly as it is, again! The type-dispatching is still taken care of by the individual transforms like you proposed in #1. 
Really the only thing that changes is this: in #1, you can pass a dict and _each_ transform reads every key, applies the appropriate transform (based on the type, not the key), and returns a dictionary. But that does not need to be the case: as far as I understand, the keys are not used in any meaningful way by the transform, only the type matters. In #2, the transform only sees the type, never the key.
I don't see why each transform of a chain should be able to read dictionaries and tensor(-like) objects. I think we can do with only the latter, and take care of the first loop over the dictionary keys at a higher level.
In short, here's the high level view of the two approaches:

#1 :
```python
for t in transform:
   for k, item in input_dict.keys(): # taken care of by t -- t sees a dictionary
      input_dict[k] = t(item) # t processes item based on its type
```

#2:
```python
for k, item in input_dict.keys(): # taken care of by TransformDispatch
   for t in transform:
      item = t(item) # t processes item based on its type -- t never sees a dictionary and is oblivious to the key
   input_dict[k] = item
```
Hope that helps!

**Comment by vmoens** [comment]
Here is an example of how that would work in #2:
```python
t = Compose(RandomHorizontalFlip(p=0.5), RandomCrop(...))
t_wrapped = TransformDispatch(t)

input_dict = {'image': Image(...), 'bbox': BoundingBox(...)}
input_dict = t_wrapped(input_dict)

t(input_dict) # Error: input must be tensor(-like) object, not dictionary
for k, item in input_dict.items():
    input_dict[k] = t(item) # runs but applies different random transformation every time
```

**Comment by pmeier** [comment]
To summarize, what you propose is to move the actual (nested) container dispatch to a separate transform `TransformDispatch`, correct? Together with the new datasets a typical usage would look like `dataset.map(TransformDispatch(transform))` whereas with the approach in #1 it would look like `dataset.map(transform)`.

So far I see two disadvantages and one possible advantage for this approach.

### Disadvantages

- The user needs to remember that a wrapper object is needed for using the transformation with containers. Not an totally unreasonable ask, but a worse UX nonetheless.
- Since now each sample "gets ripped apart" we need extra handling, and thus extra opportunities for mistakes, for transforming all parts of the sample with the same parameters. Again, the added complexity is not off the charts, but it is still there.

### (Possible) Advantages

- With the proposed approach here, a transform should be scriptable. I label this as a possible advantage, since the functional interface described in #1 should offer this to. Still, your approach would have the better UX, since the transform could be scripted directly without scripting a class method of it.

### Assessment

Since both approaches actually achieve the same, I think it comes down to the question to which needs we cater more: users who want to use our transforms for training and experimenting or users who want to put the transformations in a production pipeline. Depending on which approach we chose, one group gets a worse UX than the other one.

I'm (naturally) leaning towards my approach, i.e. a better UX for researchers and users who train models, but this is not for me to decide.

File: transforms_poc.py [hunk 2]
```diff
@@ -10,24 +10,24 @@
...
+sample_a = dispatch_transform(dict(image=image, bbox=bbox))
 torch.testing.assert_close(sample_a["image"], image_e)
 torch.testing.assert_close(sample_a["bbox"].convert("xyxy"), bbox_e)
 
-image_a, bbox_a = transform(image, bbox)
-torch.testing.assert_close(image_a, image_e)
-torch.testing.assert_close(bbox_a.convert("xyxy"), bbox_e)
-
-composed_transform = transforms.Compose(transform, transform)
+# FIXME: tuple dispatch is not yet supported
```
Comment (pmeier) [inline transforms_poc.py]: Here.

File: transforms_poc.py [hunk 3]
```diff
@@ -10,24 +10,24 @@
...
+# image_a, bbox_a = dispatch_transform(image, bbox)
+# torch.testing.assert_close(image_a, image_e)
+# torch.testing.assert_close(bbox_a.convert("xyxy"), bbox_e)
 
-image_a, bbox_a = composed_transform(image, bbox)
-torch.testing.assert_close(image_a, image)
-torch.testing.assert_close(bbox_a.convert("xyxy"), bbox)
+# FIXME: composed transforms are not yet supported
```
Comment (pmeier) [inline transforms_poc.py]: And here.

**Review by pmeier — COMMENTED** [review]
@vmoens I've adapted the PoC script to fit your implementation. There are two cases left where the transform implementation needs to be adapted.

**Comment by pmeier** [comment]
@datumbox, @fmassa, @NicolasHug, @prabhat00155 you should read #1 first, since this PR is proposing changes based on that.

**Comment by vmoens** [comment]
> ### Disadvantages
> * The user needs to remember that a wrapper object is needed for using the transformation with containers. Not an totally unreasonable ask, but a worse UX nonetheless.

Why is that worse? In #1 the user still needs to remember that he has one class that can do multiple things with multiple inputs, and return different output types depending on the input type, which is not a trivial thing in many implementations (like those matplotlib functions that can return integers or np arrays -- it can get really messy when you are coding things that expect one and only one output type). Here, the users knows that whatever they do, the input and output types are defined by the class they're using. From a UX perspective i find it way better personally. For instance, if you were to re-use a code snippet, you'd be sure of the behaviour of each bit of that code as there would be no surprise on the output type of the transform.
Also, and still from a UX perspective, the help headers of #1 would quite cryptic IMHO. Smth like "This class takes either XX or YY or ZZ as input and returns the same type with randomness controlled by doing blabla" is way more complicated than: class A: "transforms a dictionary of inputs with a defined transform, takes care of randomness" / class B: "transforms a tensor-like object by doing blabla". It's a single-class / single-responsiblity approach. 

> * Since now each sample "gets ripped apart" we need extra handling, and thus extra opportunities for mistakes, for transforming all parts of the sample with the same parameters. Again, the added complexity is not off the charts, but it is still there.

Here too i don't necessarily see it that way. The functionality is the same, the opportunity for mistakes is roughly equal. What kind of mistakes do you have in mind, that would occur in #2 and not in #1?

As an advantage, i would also add that from a testing perspective, #1 would require to test that transforms:
1.  that take a tensor(-like) object return the same type
2.  that take a certain data structure (e.g. `dict`) return the same data structure
3.  do what they should on each data type
4.  control randomness appropriately for multiple inputs

That's a lot of responsibility fo a single class. #2 assigns (2) and (4) to a separate wrapper. 
Finally, still from a UX perspective, having to understand the what a transform does in #1 by reading the code is more complex than in #2 (multiple `apply` functions that call each other with various types etc.) It is hard so see what is going on.

File: torchvision/transforms/_core.py [hunk 4]
```diff
@@ -23,17 +24,40 @@ def query_sample(sample: Any, fn: Callable[[Any], Optional[T]]) -> Iterator[T]:
...
+
 class Transform(nn.Module):
     _KNOWN_FEATURE_TYPES = {
         "_".join([part.lower() for part in re.findall("[A-Z][^A-Z]*", feature_type.__name__)]): feature_type
         for feature_type in (Image, BoundingBox)
     }
 
     def __init_subclass__(cls, *, auto_register: bool = True):
+        cls.initialized = False
+        cls.reset_auto = True
```
Comment (pmeier) [inline torchvision/transforms/_core.py]: These attributes don't belong here, as they would be the same on every `Transform`. Since they should vary on a per-instance base, the need to go into `__init__`.

File: torchvision/transforms/_core.py [hunk 5]
```diff
@@ -23,17 +24,40 @@ def query_sample(sample: Any, fn: Callable[[Any], Optional[T]]) -> Iterator[T]:
...
         cls._feature_transforms: Dict[Type[Feature], Callable] = {}
         if auto_register:
             cls._auto_register()
 
+    def set_reset_auto(self, mode=True):
+        self.reset_auto = mode
+
+    def wipeout_(self):
```
Comment (pmeier) [inline torchvision/transforms/_core.py:65]: Shouldn't these two methods be private?

File: torchvision/transforms/_core.py [hunk 6]
```diff
@@ -23,17 +24,40 @@ def query_sample(sample: Any, fn: Callable[[Any], Optional[T]]) -> Iterator[T]:
             yield result
 
 
+class TransformDispatch:
```
Comment (pmeier) [inline torchvision/transforms/_core.py]: IMO something like `JointTransform` implies its use case better.

```suggestion
class JointTransform:
```

File: torchvision/transforms/_core.py [hunk 7]
```diff
@@ -23,17 +24,40 @@ def query_sample(sample: Any, fn: Callable[[Any], Optional[T]]) -> Iterator[T]:
...
+        self.transform = transform
+        self.transform.set_reset_auto(False)
+
+    def __call__(self, inputs):
+        if isinstance(inputs, collections.abc.Mapping):
+            inputs = {key: self.transform(value) for key, value in inputs.items()}
+        elif isinstance(inputs, collections.abc.Sequence):
+            inputs = tuple(self.transform(input) for input in inputs)
+        self.transform.wipeout_()
+        return inputs
```
Comment (pmeier) [inline torchvision/transforms/_core.py]: 1. This should be callable with a variable number of `inputs`, i.e. `transform(image, bbox)`.
2. The inputs might be nested containers, so IMO recursing through it is the easiest solution.

**Review by pmeier — COMMENTED** [review]
> Why is that worse? In #1 the user still needs to remember that he has one class that can do multiple things with multiple inputs, and return different output types depending on the input type, which is not a trivial thing in many implementations (like those matplotlib functions that can return integers or np arrays -- it can get really messy when you are coding things that expect one and only one output type)

I think this is misrepresenting the approach in #1. Yes, the output type is dependent on the input type, but in a one-to-one fashion: Put in a tensor, get back a tensor. Put in a dictionary of tensors, get back a dictionary of tensors. Put in some twisted nested container, get back the same.

> Here, the users knows that whatever they do, the input and output types are defined by the class they're using.

I fail to see how this is dependent on the transform class. Each transform will still handle multiple features but just not containers, right?

> For instance, if you were to re-use a code snippet, you'd be sure of the behaviour of each bit of that code as there would be no surprise on the output type of the transform.

Could you give an example were this would be an issue? If you reuse your transform and now put in a container instead of a tensor, what would you expect it to do?

> Also, and still from a UX perspective, the help headers of #1 would quite cryptic IMHO. Smth like "This class takes either XX or YY or ZZ as input and returns the same type with randomness controlled by doing blabla" is way more complicated than: class A: "transforms a dictionary of inputs with a defined transform, takes care of randomness" / class B: "transforms a tensor-like object by doing blabla". It's a single-class / single-responsiblity approach.

+1 for single-class / single-responsiblity. I don't see the issue for the documentation though. Due to one-to-one relation of input and output, the documentation can be straight forward. 

> Here too i don't necessarily see it that way. The functionality is the same, the opportunity for mistakes is roughly equal. What kind of mistakes do you have in mind, that would occur in #2 and not in #1?

After a second read, I maybe need to scratch that. Is there a use case for `wipe_out_` and `set_reset_auto` being public? If yes, this makes the interface more complicated, since something like this is not needed in #1. If they can be made private (including all the related attributes), the transforms have the same UI, rendering my point invalid.

> As an advantage, i would also add that from a testing perspective, #1 would require to test that transforms:
> 
>     1. that take a tensor(-like) object return the same type
> 
>     2. that take a certain data structure (e.g. `dict`) return the same data structure
> 
>     3. do what they should on each data type
> 
>     4. control randomness appropriately for multiple inputs
> 
> That's a lot of responsibility fo a single class. #2 assigns (2) and (4) to a separate wrapper.

True, but not hard to solve. Apart from 3. this can be done by automated testing that looks the same for every transformation.

> Finally, still from a UX perspective, having to understand the what a transform does in #1 by reading the code is more complex than in #2 (multiple apply functions that call each other with various types etc.) It is hard so see what is going on.

Well, #2 currently doesn't handle nested containers, which it needs to. I'd wager you'll see the very same patterns here if the features are on par.

File: torchvision/transforms/_core.py (hunk 5 above)
Comment (vmoens) [reply torchvision/transforms/_core.py:65]: not sure, i can imagine that people will want to play with that in some cases.

File: torchvision/transforms/_core.py (hunk 7 above)
Comment (vmoens) [reply torchvision/transforms/_core.py]: > This should be callable with a variable number of inputs, i.e. transform(image, bbox)

Is that necessary? I don't really see the point, a loop can do the job. Again, IMO it is weird to have a function that can take a dictionary _and_ any length of args. What happens if it gets both?

> The inputs might be nested containers, so IMO recursing through it is the easiest solution.

I think that there is a design choice that needs to be made here: is it important that anything can happen and that we take any possibility into account, or do we want to constrain the usage of our classes to a specific use case?
Are there datasets that have those complex composite structures, like `Tuple[Dict[str, Tensor], Tuple[Tensor, ...], Tensor]`? If that can happen i'm ok with recursive of course.

Comment (pmeier) [reply torchvision/transforms/_core.py]: > Are there datasets that have those complex composite structures

Yes. COCO for example looks like `dict(image=Image(...), annotations=[dict(bbox=BoundingBox(...)), ...])`. So you have a dict in a list in a dict. There might be even crazier constructs for datasets that we currently don't support. At this point there is no reason not to support parsing "all" nested containers. 

> > This should be callable with a variable number of inputs, i.e. transform(image, bbox)
> 
> Is that necessary? I don't really see the point, a loop can do the job. 

No, not necessary, but makes a good UX. "If the transform can parse single tensors and containers at the same time, why do I need to pass in two tensors in a container?"

File: torchvision/transforms/_core.py (hunk 5 above)
Comment (pmeier) [reply torchvision/transforms/_core.py:65]: Can you give examples? Because if there is a need to use these outside of the `JointTransform`, your porposal is more complex / error-prone, because #1 doesn't need these methods and attributes.

Comment (vmoens) [reply torchvision/transforms/_core.py:65]: Frankly it is a rather minor point, we can make them private np

File: torchvision/transforms/_core.py (hunk 7 above)
Comment (vmoens) [reply torchvision/transforms/_core.py]: Ok agreed then, we can check if it is a tensor-like and pass that to transforms, otherwise call JointTransforms again.
But then in your example why do we need to unpack the list?

Wouldn't that work?

```python
class JointTransform:
    def __init__(self, transform):
        self.transform = transform
        self.transform.set_reset_auto(False)

    def __call__(self, inputs, _top=True):
        if isinstance(inputs, collections.abc.Mapping):
            inputs = {key: self(value, _top=False) for key, value in inputs.items()}
        elif isinstance(inputs, collections.abc.Sequence):
            inputs = tuple(self(input, _top=False) for input in inputs)
        else:
            inputs = self.transform(inputs)
        if _top:
            self.transform.wipeout_()
        return inputs
```

Why do we need to account for `__call__(self, *inputs)`?

Comment (pmeier) [reply torchvision/transforms/_core.py]: Again, we don't, but it makes a good UX. If I want to transform an `image` and a `bounding_box`, why do I need to put them in a container first? This is just for convenience since `image, bbox = transform(image, bbox)` is nicer than `image, bbox = transform((image, bbox))`. 

Still, let's keep the discussion narrow in this PR. Adding this would not be an issue here. If you want to discuss this further (not saying you shouldn't), please do so in #1, where I added this to the list of required features the new API should support.

---

By moving the recursive call outside of `__call__` you can avoid the `_top` parameter all together, making it more clear:

```python
class JointTransform:
    def __init__(self, transform):
        self.transform = transform
        self.transform.set_reset_auto(False)

    def _recurse(self, inputs):
        if isinstance(inputs, collections.abc.Mapping):
            return {key: self._recurse(value) for key, value in inputs.items()}
        elif isinstance(inputs, collections.abc.Sequence):
            return [self._recurse(input) for input in inputs]
        else:
            return self.transform(inputs)

    def __call__(self, inputs):
        outputs = self._recurse(inputs)
        self.transform.wipeout_()
        return outputs
```

**Comment by pmeier** [comment]
@vmoens Can we close this in the light of https://github.com/pmeier/torchvision-datasets-rework/pull/1#discussion_r699298253? Although in the end grew fond of having the container dispatch outside of the transform, if the parameter sampling depends on more than one feature, it seems a lot easier having everything in one class.

**Comment by vmoens** [comment]
I agree let's close it
